core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -333,6 +347,122 @@ public long nextFetchOffset() {
         }
     }
 
+    void updateOffsetsOnLsoMovement() {
+        long logStartOffset = offsetForEarliestTimestamp();
+        lock.writeLock().lock();
+        try {
+            if (logStartOffset > startOffset) {
+                log.debug("Updating start offset for share partition: {}-{} from: {} to: {} since LSO has moved to: {}",
+                        groupId, topicIdPartition, startOffset, logStartOffset, logStartOffset);
+                if (cachedState.isEmpty()) {
+                    startOffset = logStartOffset;
+                    endOffset = logStartOffset;
+                } else {
+                    List<PersisterStateBatch> stateBatches = new ArrayList<>();
+                    for (Map.Entry<Long, InFlightBatch> entry : cachedState.entrySet()) {
+                        long batchStartOffset = entry.getKey();
+                        // We do not need to transition state of batches/offsets that are later than the new log start offset.
+                        if (batchStartOffset >= logStartOffset) {
+                            break;
+                        }
+                        InFlightBatch inFlightBatch = entry.getValue();
+                        boolean fullMatch = inFlightBatch.firstOffset >= startOffset && inFlightBatch.lastOffset < logStartOffset;
+
+                        // Maintain state per offset if the inflight batch is not a full match or the offset state is managed.
+                        if (!fullMatch || inFlightBatch.offsetState != null) {
+                            log.debug("Subset or offset tracked batch record found while trying to update offsets and cached" +
+                                            " state map due to LSO movement, batch: {}, offsets to update - " +
+                                            "first: {}, last: {} for the share partition: {}-{}", inFlightBatch, startOffset,
+                                    logStartOffset - 1, groupId, topicIdPartition);
+
+                            if (inFlightBatch.offsetState == null) {
+                                if (inFlightBatch.batchState() != RecordState.AVAILABLE) {
+                                    continue;
+                                }
+                                inFlightBatch.maybeInitializeOffsetStateUpdate();
+                            }
+
+                            archivePerOffsetBatchRecords(inFlightBatch, startOffset, logStartOffset - 1, stateBatches);
+                        } else {
+                            // The in-flight batch is a full match hence change the state of the complete batch.
+                            archiveCompleteBatch(inFlightBatch, stateBatches);
+                        }
+                    }
```

Comment:
Can we divide the method with subset of the code. Methods are generally long in this class.